subsys/logging/log_backend_rpc.c

```
@@ -0,0 +1,310 @@
+		return;
+	}
+
+	crash_log_max_size = (size_t)rc - sizeof(crash_log_header_t);
+}
+
+static void retain(const uint8_t *data, size_t length)
+{
+	const struct device *const crash_log = CRASH_LOG_DEV;
+	crash_log_header_t header;
+	int rc;
+
+	if (crash_log_cur_size >= crash_log_max_size) {
+		/* Crash log partition overflow. */
+		return;
+	}
+
+	length = MIN(length, crash_log_max_size - crash_log_cur_size);
+
+	rc = retention_write(crash_log, sizeof(header) + crash_log_cur_size, data, length);
+
+	if (rc) {
+		return;
+	}
+
+	crash_log_cur_size += length;
+	header.size = crash_log_cur_size;
+
+	(void)retention_write(crash_log, 0, (const uint8_t *)&header, sizeof(header));
+}
+
+static void log_rpc_get_crash_log_handler(const struct nrf_rpc_group *group,
+					  struct nrf_rpc_cbor_ctx *ctx, void *handler_data)
+{
+	const struct device *const crash_log = CRASH_LOG_DEV;
+	size_t offset;
+	size_t length;
+	bool decoded_ok;
+	crash_log_header_t header;
+	struct nrf_rpc_cbor_ctx rsp_ctx;
+
+	decoded_ok = zcbor_uint_decode(ctx->zs, &offset, sizeof(offset));
+	decoded_ok = decoded_ok && zcbor_uint_decode(ctx->zs, &length, sizeof(length));
+	nrf_rpc_cbor_decoding_done(group, ctx);
+
+	if (!decoded_ok) {
+		goto err;
+	}
+
+	if (retention_is_valid(crash_log) != 1) {
+		goto err;
+	}
+
+	if (retention_read(crash_log, 0, (uint8_t *)&header, sizeof(header))) {
+		goto err;
+	}
+
+	if (offset >= header.size) {
+		goto err;
+	}
+
+	length = MIN(length, header.size - offset);
+	NRF_RPC_CBOR_ALLOC(group, rsp_ctx, 5 + length);
+
+	if (!zcbor_bstr_start_encode(rsp_ctx.zs)) {
+		goto discard;
+	}
+
+	if (retention_read(crash_log, sizeof(header) + offset, rsp_ctx.zs[0].payload_mut, length)) {
+		goto discard;
+	}
+
+	rsp_ctx.zs[0].payload_mut += length;
+
+	if (!zcbor_bstr_end_encode(rsp_ctx.zs, NULL)) {
+		goto discard;
+	}
+
+	goto send;
+
+discard:
+	NRF_RPC_CBOR_DISCARD(group, rsp_ctx);
+err:
+	/* Send empty string */
```

Why is the empty string required?

---

Because the API contract which I documented is that an empty string means "end of file". We could return an error code as well but for a caller it doesn't really make a difference why the crash log is not accessible (I thought).